pkg/selfupdate/selfupdate.go

```diff
@@ -0,0 +1,121 @@
+package selfupdate
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/blang/semver/v4"
+	"github.com/google/go-github/github"
+	"github.com/minio/selfupdate"
+)
+
+type SelfUpdater struct {
+	assetId     int64
+	githubOwner string
+	githubRepo  string
+	assetUrl    string
+	Version     semver.Version
+}
+
+func NewSelfUpdater(ctx context.Context, githubOwner, githubRepo string) (*SelfUpdater, error) {
+	var err error
+	var githubRelease *github.RepositoryRelease
+
+	selfUpdater := new(SelfUpdater)
+	selfUpdater.githubOwner = githubOwner
+	selfUpdater.githubRepo = githubRepo
+
+	client := github.NewClient(nil)
+	if githubRelease, _, err = client.Repositories.GetLatestRelease(ctx, githubOwner, githubRepo); err != nil {
+		return nil, err
+	}
+	if err = selfUpdater.fetchVersion(githubRelease); err != nil {
+		return nil, err
+	}
+	if err = selfUpdater.fetchAsset(ctx, client, githubRelease); err != nil {
+		return nil, err
+	}
+	return selfUpdater, err
+}
+
+func (selfUpdater *SelfUpdater) fetchVersion(githubRelease *github.RepositoryRelease) error {
+	var err error
+
+	selfUpdater.Version, err = semver.ParseTolerant(githubRelease.GetTagName())
+	return err
+}
+
+func (selfUpdater *SelfUpdater) fetchAsset(ctx context.Context, client *github.Client, githubRelease *github.RepositoryRelease) error {
+	var err error
+
+	assetSuffix := fmt.Sprintf("%s_%s.tar.gz", runtime.GOOS, runtime.GOARCH)
+	for _, asset := range githubRelease.Assets {
+		if strings.HasSuffix(asset.GetName(), assetSuffix) {
+			selfUpdater.assetId = asset.GetID()
+			_, selfUpdater.assetUrl, err = client.Repositories.DownloadReleaseAsset(ctx, selfUpdater.githubOwner, selfUpdater.githubRepo, selfUpdater.assetId)
```

Comment:
same comment -  on err its unclear what is the state of assetUrl
I agree its really clean that way but explicitly assigning when its clear err != nil is friendlier for other developers